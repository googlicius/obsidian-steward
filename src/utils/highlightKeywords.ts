interface HighlightOptions {
	beforeMark: string;
	afterMark: string;
	contextChars: number;
	minMatchLength: number;
}

export function highlightKeywords(
	keywords: string[],
	content: string,
	options: Partial<HighlightOptions> = {}
): string[] {
	// Default options
	const defaultOptions: HighlightOptions = {
		beforeMark: '==',
		afterMark: '==',
		contextChars: 50,
		minMatchLength: 3,
	};

	const opts = { ...defaultOptions, ...options };

	// Filter out empty keywords
	const filteredKeywords = keywords.filter(keyword => keyword.trim().length > 0);

	if (filteredKeywords.length === 0 || !content) {
		return [];
	}

	// Generate all possible n-grams from the keywords
	const ngrams = generateNgrams(filteredKeywords);

	// Find all matches in the content
	const matches = findMatches(content, ngrams);

	// Sort matches by length (longer matches take precedence)
	matches.sort((a, b) => {
		// Sort by length (descending)
		if (b.text.length !== a.text.length) {
			return b.text.length - a.text.length;
		}
		// If same length, sort by start position
		return a.index - b.index;
	});

	// Remove overlapping matches
	const nonOverlappingMatches = removeOverlappingMatches(matches);

	// Generate highlighted results
	return generateHighlightedResults(content, nonOverlappingMatches, opts);
}

interface Match {
	text: string;
	index: number;
}

// Export for testing purposes
export function generateNgrams(keywords: string[]): string[] {
	const ngrams: Set<string> = new Set();

	for (const keyword of keywords) {
		// Add the full keyword
		ngrams.add(keyword);

		// Generate smaller n-grams for keywords with multiple words
		const words = keyword.split(/\s+/).filter(word => word.length > 0);
		if (words.length > 1) {
			// Add individual words (1-grams)
			for (const word of words) {
				if (word.length > 0) {
					ngrams.add(word);
				}
			}

			// Generate all possible n-grams of lengths 2 to words.length-1
			for (let n = 2; n < words.length; n++) {
				// For each possible starting position
				for (let i = 0; i <= words.length - n; i++) {
					const nGram = words.slice(i, i + n).join(' ');
					ngrams.add(nGram);
				}
			}
		}
	}

	// Sort ngrams by length (descending) to prioritize longer matches
	return Array.from(ngrams).sort((a, b) => b.length - a.length);
}

function findMatches(content: string, terms: string[]): Match[] {
	const matches: Match[] = [];
	const lowerContent = content.toLowerCase();

	for (const term of terms) {
		const lowerTerm = term.toLowerCase();
		let index = 0;

		while ((index = lowerContent.indexOf(lowerTerm, index)) !== -1) {
			matches.push({
				text: content.substring(index, index + term.length),
				index,
			});
			index += 1; // Move forward to find next occurrence
		}
	}

	return matches;
}

function removeOverlappingMatches(matches: Match[]): Match[] {
	if (matches.length <= 1) return matches;

	const result: Match[] = [];
	let lastEnd = -1;

	for (const match of matches) {
		const currentEnd = match.index + match.text.length;

		if (match.index > lastEnd) {
			result.push(match);
			lastEnd = currentEnd;
		}
	}

	return result;
}

function generateHighlightedResults(
	content: string,
	matches: Match[],
	options: HighlightOptions
): string[] {
	if (matches.length === 0) return [];

	const highlights: string[] = [];

	for (const match of matches) {
		const matchText = match.text;
		const matchIndex = match.index;

		// Calculate start and end of context
		const contextStart = Math.max(0, matchIndex - options.contextChars);
		const contextEnd = Math.min(
			content.length,
			matchIndex + matchText.length + options.contextChars
		);

		// Get context with highlight
		const before = content.substring(contextStart, matchIndex);
		const highlighted = options.beforeMark + matchText + options.afterMark;
		const after = content.substring(matchIndex + matchText.length, contextEnd);

		const result = before + highlighted + after;
		highlights.push(result);
	}

	return highlights;
}
