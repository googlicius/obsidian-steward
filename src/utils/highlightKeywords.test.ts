import { generateNgrams } from './highlightKeywords'; // Adjust import path as needed

describe('generateNgrams', () => {
  it('should handle empty array', () => {
    const keyword = '';
    const result = generateNgrams(keyword);
    expect(result).toEqual([]);
  });

  it('should handle single-word keyword', () => {
    const keyword = 'test';
    const result = generateNgrams(keyword);
    expect(result).toEqual(['test']);
  });

  it('should trim whitespace from the keyword', () => {
    const keyword = '  test  ';
    const result = generateNgrams(keyword);
    expect(result).toEqual(['test']);
  });

  it('should generate all possible n-grams for a multi-word keyword', () => {
    const keyword = 'hello world';
    const result = generateNgrams(keyword);

    expect(result).toEqual(['hello world']);
  });

  it('should generate n-grams that are at least 70% of the keyword length', () => {
    const keyword = 'one two three four five six';
    const result = generateNgrams(keyword);

    expect(result).toEqual([
      'one two three four five six',
      'one two three four five',
      'two three four five six',
      'two three four five',
      'three four five six',
      'one two three four',
    ]);
  });
});
