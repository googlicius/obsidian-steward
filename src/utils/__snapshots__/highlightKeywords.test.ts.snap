// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateNgrams should correctly handle five-word keyword 1`] = `
[
  "one two three four five",
  "two three four five",
  "one two three four",
  "three four five",
  "two three four",
  "one two three",
  "three four",
  "two three",
  "four five",
  "one two",
  "three",
  "four",
  "five",
  "one",
  "two",
]
`;

exports[`highlightKeywords should highlight keywords in multi-line content 1`] = `
[
  "This is a ==test content==.

With multiple lines.

And test content repeated",
  "his is a test content.

With multiple lines.

And ==test content== repeated.",
]
`;

exports[`highlightKeywords should highlight keywords in multi-line content 2 1`] = `
[
  "s for confirmation. That part works now.

Another ==part checks the total number== of files; if there are no files, it will not ask ",
  "Another part checks the total number of files; if ==the==re are no files, it will not ask users for confirm",
  "iles, it will not ask users for confirmation when ==the== destination does not exist. However, there is a r",
  "ion when the destination does not exist. However, ==the==re is a redundant query for the files occurring tw",
  "ot exist. However, there is a redundant query for ==the== files occurring twice: once for checking file exi",
]
`;
